File: tests/python/test_sfile.py
import gc
import os
import sys
from pathlib import Path

import numpy as np
import pytest

import eqmdsk


def _write(path: Path, content: bytes) -> Path:
    path.write_bytes(content)
    return path


def test_sfile_without_titles(tmp_path):
    source = _write(
        tmp_path / "s.no-title", b"1 2 0.1 0.2\n3 4 0.3 0.4\n"
    )
    sfile = eqmdsk.SFile(source)

    assert "XLABEL" not in sfile
    assert "YLABEL" not in sfile
    assert "TITLE" not in sfile
    np.testing.assert_array_equal(sfile["X"], [1.0, 3.0])
    np.testing.assert_array_equal(sfile["Y"], [2.0, 4.0])
    np.testing.assert_array_equal(sfile["DX"], [0.1, 0.3])
    np.testing.assert_array_equal(sfile["DY"], [0.2, 0.4])


def test_sfile_empty_and_titles_only(tmp_path):
    empty = _write(tmp_path / "s.empty", b"")
    sfile = eqmdsk.SFile(empty)
    assert sfile["X"].shape == (0,)
    sfile.write()
    assert empty.read_bytes() == b""

    titled = _write(tmp_path / "s.titled", b"x\ry\rtitle\r")
    sfile = eqmdsk.SFile(titled)
    assert sfile["XLABEL"] == "x"
    assert sfile["YLABEL"] == "y"
    assert sfile["TITLE"] == "title"
    assert sfile["X"].shape == (0,)
    output = tmp_path / "s.titled.output"
    sfile.write(output)
    assert eqmdsk.SFile(output)["TITLE"] == "title"


def test_sfile_three_titles_and_fortran_exponents(tmp_path):
    source = _write(
        tmp_path / "s.three-title",
        b"major radius\r\npressure\r\nFIT TITLE\r\n"
        b"1.0D+00 2.0d+00 3.0E-1 4.0e-1\r\n",
    )
    sfile = eqmdsk.SFile(source)

    assert sfile["XLABEL"] == "major radius"
    assert sfile["YLABEL"] == "pressure"
    assert sfile["TITLE"] == "FIT TITLE"
    np.testing.assert_array_equal(sfile["X"], [1.0])
    np.testing.assert_array_equal(sfile["DY"], [0.4])


def test_sfile_normalizes_interstitial_text_on_write(tmp_path):
    source = _write(
        tmp_path / "s.extra",
        b"x label\ny label\nfit title\n"
        b"1 10 0.1 1\n"
        b"COMMENT BETWEEN\r\n"
        b"2024-run metadata\n"
        b"2 20 0.2 2\n"
        b"COMMENT AFTER\0OPAQUE",
    )
    target = tmp_path / "s.extra.roundtrip"

    sfile = eqmdsk.SFile(source)
    sfile.write(target)

    output = target.read_bytes()
    assert b"COMMENT BETWEEN" not in output
    assert b"2024-run metadata" not in output
    assert b"COMMENT AFTER" not in output
    reparsed = eqmdsk.SFile(target)
    np.testing.assert_array_equal(reparsed["X"], sfile["X"])


@pytest.mark.parametrize(
    "row",
    [
        b"1 2 3\n",
        b"1 2 3 4 5\n",
        b"1 2 broken 4\n",
    ],
)
def test_sfile_rejects_bad_data_columns(tmp_path, row):
    source = _write(tmp_path / "s.bad-columns", row)
    with pytest.raises(eqmdsk.ParseError) as caught:
        eqmdsk.SFile(source)
    assert caught.value.filename == str(source)
    assert caught.value.line == 1
    assert caught.value.column == 1


@pytest.mark.parametrize("value", ["1e999", "1e-999"])
def test_sfile_rejects_unrepresentable_data_values(tmp_path, value):
    source = _write(
        tmp_path / "s.unrepresentable",
        f"{value} 2 3 4\n".encode("ascii"),
    )
    with pytest.raises(eqmdsk.ParseError, match="finite and representable"):
        eqmdsk.SFile(source)


def test_sfile_rejects_strict_numeric_mixed_row_after_data(tmp_path):
    source = _write(
        tmp_path / "s.bad-row-after-data",
        b"0 0 0 0\n1 damaged metadata\n",
    )
    with pytest.raises(eqmdsk.ParseError) as caught:
        eqmdsk.SFile(source)
    assert caught.value.line == 2
    assert caught.value.column == 1


def test_sfile_reports_unicode_diagnostic_paths(tmp_path):
    source = _write(tmp_path / "s-坏文件-平衡", b"1 2 broken 4\n")
    with pytest.raises(eqmdsk.ParseError) as caught:
        eqmdsk.SFile(source)
    assert caught.value.filename == str(source)
    assert str(source) in str(caught.value)

    missing = tmp_path / "s-不存在-平衡"
    with pytest.raises(eqmdsk.IOError) as caught:
        eqmdsk.SFile(missing)
    assert str(missing) in str(caught.value)


@pytest.mark.skipif(
    os.name == "nt" or sys.platform == "darwin",
    reason="invalid UTF-8 filesystem paths are unavailable on this platform",
)
def test_invalid_utf8_path_does_not_break_exception_translation(tmp_path):
    source = tmp_path / os.fsdecode(b"s-invalid-\xff")
    source.write_bytes(b"1 2 broken 4\n")

    with pytest.raises(eqmdsk.ParseError) as caught:
        eqmdsk.SFile(source)
    assert "\ufffd" in caught.value.filename
    assert "\ufffd" in str(caught.value)

    missing = tmp_path / os.fsdecode(b"s-missing-\xff")
    with pytest.raises(eqmdsk.IOError) as caught:
        eqmdsk.SFile(missing)
    assert "\ufffd" in str(caught.value)


def test_sfile_max_digits_precision_roundtrip(tmp_path):
    source = _write(tmp_path / "s.precision", b"0 0 0 0\n")
    target = tmp_path / "s.precision.output"
    sfile = eqmdsk.SFile(source)
    values = np.array(
        [
            np.nextafter(1.0, 2.0),
            -np.finfo(np.float64).tiny,
            np.finfo(np.float64).max,
            np.nextafter(0.0, 1.0),
        ],
        dtype=np.float64,
    )
    sfile["X"][0], sfile["Y"][0], sfile["DX"][0], sfile["DY"][0] = values
    sfile.write(target)

    reparsed = eqmdsk.SFile(target)
    actual = np.array(
        [reparsed["X"][0], reparsed["Y"][0], reparsed["DX"][0], reparsed["DY"][0]]
    )
    np.testing.assert_array_equal(
        actual.view(np.uint64), values.view(np.uint64)
    )


def test_sfile_arrays_are_writable_and_write_defaults_to_source(tmp_path):
    source = _write(tmp_path / "s.modify", b"1 2 0.1 0.2\n")
    sfile = eqmdsk.SFile(source)

    assert sfile["Y"].flags.writeable
    view = sfile["Y"]
    assert view.dtype == np.float64
    assert view.shape == (1,)
    assert view.strides == (8,)
    assert view.flags.c_contiguous
    pointer = view.__array_interface__["data"][0]
    view[0] = 42.25
    sfile.write()
    assert eqmdsk.SFile(source)["Y"][0] == 42.25
    del sfile
    gc.collect()
    assert view.__array_interface__["data"][0] == pointer
    assert view[0] == 42.25


def test_sfile_validates_equal_lengths_and_finite_values(tmp_path):
    source = _write(tmp_path / "s.validation", b"1 2 3 4\n")
    sfile = eqmdsk.SFile(source)
    with pytest.raises(ValueError, match="preserve the existing array length"):
        sfile["DX"] = np.array([1.0, 2.0])

    sfile = eqmdsk.SFile(source)
    sfile["DY"][0] = np.inf
    with pytest.raises(eqmdsk.ValidationError):
        sfile.write(tmp_path / "nonfinite")


@pytest.mark.skipif(not Path("/dev/full").exists(), reason="requires /dev/full")
def test_sfile_reports_deferred_write_failure(tmp_path):
    source = _write(tmp_path / "s.write-error", b"1 2 3 4\n")
    sfile = eqmdsk.SFile(source)
    with pytest.raises(eqmdsk.IOError):
        sfile.write(Path("/dev/full"))

File: tests/cpp/test_sfile.cpp
#include <cassert>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <limits>
#include <sstream>
#include <string>

#include "eqmdsk/error.hpp"
#include "eqmdsk/field.hpp"
#include "eqmdsk/sfile.hpp"

namespace {

class TemporaryDirectory {
 public:
  TemporaryDirectory() {
    const auto stamp = std::chrono::high_resolution_clock::now()
                           .time_since_epoch()
                           .count();
    path_ = std::filesystem::temp_directory_path() /
            ("eqmdsk-sfile-test-" + std::to_string(stamp));
    std::filesystem::create_directories(path_);
  }

  ~TemporaryDirectory() { std::filesystem::remove_all(path_); }

  const std::filesystem::path& path() const noexcept { return path_; }

 private:
  std::filesystem::path path_;
};

void write_bytes(const std::filesystem::path& path, const std::string& bytes) {
  std::ofstream output(path, std::ios::binary | std::ios::trunc);
  output.write(bytes.data(), static_cast<std::streamsize>(bytes.size()));
  assert(output.good());
}

std::string read_bytes(const std::filesystem::path& path) {
  std::ifstream input(path, std::ios::binary);
  std::ostringstream output;
  output << input.rdbuf();
  return output.str();
}

const eqmdsk::DoubleVector& vector(const eqmdsk::SFile& file,
                                   const char* name) {
  return std::get<eqmdsk::DoubleVector>(file.at(name));
}

eqmdsk::DoubleVector& vector(eqmdsk::SFile& file, const char* name) {
  return std::get<eqmdsk::DoubleVector>(file.at(name));
}

template <typename Exception, typename Function>
void assert_throws(Function&& function) {
  bool threw = false;
  try {
    function();
  } catch (const Exception&) {
    threw = true;
  }
  assert(threw);
}

void test_without_titles(const std::filesystem::path& directory) {
  const auto path = directory / "s-no-title";
  write_bytes(path, "1 2 0.1 0.2\n3 4 0.3 0.4\n");
  const eqmdsk::SFile file(path);

  assert(std::string(file.format_name()) == "SFile");
  assert(!file.fields().contains("XLABEL"));
  assert(!file.fields().contains("YLABEL"));
  assert(!file.fields().contains("TITLE"));
  assert(vector(file, "X").size() == 2);
  assert(vector(file, "X")[0] == 1.0);
  assert(vector(file, "Y")[1] == 4.0);
  assert(vector(file, "DX")[1] == 0.3);
  assert(vector(file, "DY")[0] == 0.2);
  assert(file.raw_sections().empty());
}

void test_three_titles_and_fortran_exponents(
    const std::filesystem::path& directory) {
  const auto path = directory / "s-three-title";
  write_bytes(path,
              "major radius\r\npressure\r\nFIT TITLE\r\n"
              "1.0D+00 2.0d+00 3.0E-1 4.0e-1\r\n");
  const eqmdsk::SFile file(path);

  assert(std::get<std::string>(file.at("XLABEL")) == "major radius");
  assert(std::get<std::string>(file.at("YLABEL")) == "pressure");
  assert(std::get<std::string>(file.at("TITLE")) == "FIT TITLE");
  assert(vector(file, "X")[0] == 1.0);
  assert(vector(file, "DY")[0] == 0.4);
}

void test_interstitial_text_and_roundtrip(
    const std::filesystem::path& directory) {
  const auto source = directory / "s-extra-source";
  const auto target = directory / "s-extra-target";
  std::string source_bytes =
      "x label\ny label\nfit title\n"
      "1 10 0.1 1\n"
      "COMMENT BETWEEN\r\n"
      "2024-run metadata\n"
      "2 20 0.2 2\n";
  constexpr char opaque_tail[] = "COMMENT AFTER\0OPAQUE";
  source_bytes.append(opaque_tail, sizeof(opaque_tail) - 1);
  write_bytes(source, source_bytes);

  const eqmdsk::SFile original(source);
  assert(original.raw_sections().size() == 3);
  assert(original.raw_sections()[0].name == "extra_text");
  assert(original.raw_sections()[0].data == "COMMENT BETWEEN\r\n");
  assert(original.raw_sections()[1].data == "2024-run metadata\n");
  assert(original.raw_sections()[2].data ==
         std::string(opaque_tail, sizeof(opaque_tail) - 1));
  original.write(target);

  const auto output = read_bytes(target);
  const auto first_row = output.find("1 10");
  const auto between = output.find("COMMENT BETWEEN");
  const auto numeric_text = output.find("2024-run metadata");
  const auto second_row = output.find("2 20");
  const auto after = output.find("COMMENT AFTER");
  assert(first_row < between && between < numeric_text &&
         numeric_text < second_row && second_row < after);

  const eqmdsk::SFile reparsed(target);
  assert(reparsed.raw_sections().size() == 3);
  assert(reparsed.raw_sections()[0].data == "COMMENT BETWEEN\r\n");
  assert(reparsed.raw_sections()[1].data == "2024-run metadata\n");
  assert(reparsed.raw_sections()[2].data ==
         std::string(opaque_tail, sizeof(opaque_tail) - 1));
  assert(vector(reparsed, "X")[1] == 2.0);
  assert(vector(reparsed, "Y")[0] == 10.0);
}

void test_bad_columns(const std::filesystem::path& directory) {
  const auto three = directory / "s-three-columns";
  const auto five = directory / "s-five-columns";
  const auto mixed = directory / "s-invalid-column";
  const auto mixed_after_data = directory / "s-invalid-after-data";
  const auto overflow = directory / "s-overflow";
  const auto underflow = directory / "s-underflow";
  write_bytes(three, "1 2 3\n");
  write_bytes(five, "1 2 3 4 5\n");
  write_bytes(mixed, "1 2 broken 4\n");
  write_bytes(mixed_after_data, "0 0 0 0\n1 damaged metadata\n");
  write_bytes(overflow, "1e999 2 3 4\n");
  write_bytes(underflow, "1e-999 2 3 4\n");

  assert_throws<eqmdsk::ParseError>([&] { eqmdsk::SFile file(three); });
  assert_throws<eqmdsk::ParseError>([&] { eqmdsk::SFile file(five); });
  assert_throws<eqmdsk::ParseError>([&] { eqmdsk::SFile file(mixed); });
  assert_throws<eqmdsk::ParseError>(
      [&] { eqmdsk::SFile file(mixed_after_data); });
  assert_throws<eqmdsk::ParseError>([&] { eqmdsk::SFile file(overflow); });
  assert_throws<eqmdsk::ParseError>([&] { eqmdsk::SFile file(underflow); });
}

void test_utf8_diagnostic_paths(const std::filesystem::path& directory) {
  const std::string unicode_name =
      "s-\xe5\x9d\x8f\xe6\x96\x87\xe4\xbb\xb6-\xe5\xb9\xb3\xe8\xa1\xa1";
  const auto bad = directory / std::filesystem::u8path(unicode_name);
  write_bytes(bad, "1 2 broken 4\n");

  bool saw_parse_error = false;
  try {
    const eqmdsk::SFile file(bad);
  } catch (const eqmdsk::ParseError& error) {
    saw_parse_error = true;
    assert(error.filename().find(unicode_name) != std::string::npos);
    assert(std::string(error.what()).find(unicode_name) != std::string::npos);
  }
  assert(saw_parse_error);

  const auto missing =
      directory / std::filesystem::u8path(unicode_name + "-missing");
  bool saw_io_error = false;
  try {
    const eqmdsk::SFile file(missing);
  } catch (const eqmdsk::IOError& error) {
    saw_io_error = true;
    assert(std::string(error.what()).find(unicode_name) != std::string::npos);
  }
  assert(saw_io_error);
}

void test_precision_modification_and_default_write(
    const std::filesystem::path& directory) {
  const auto path = directory / "s-precision";
  write_bytes(path, "0 0 0 0\n");
  eqmdsk::SFile file(path);

  const double values[]{
      std::nextafter(1.0, 2.0),
      -std::numeric_limits<double>::min(),
      std::numeric_limits<double>::max(),
      std::numeric_limits<double>::denorm_min(),
  };
  vector(file, "X")[0] = values[0];
  vector(file, "Y")[0] = values[1];
  vector(file, "DX")[0] = values[2];
  vector(file, "DY")[0] = values[3];
  file.write();

  eqmdsk::SFile reparsed(path);
  assert(vector(reparsed, "X")[0] == values[0]);
  assert(vector(reparsed, "Y")[0] == values[1]);
  assert(vector(reparsed, "DX")[0] == values[2]);
  assert(vector(reparsed, "DY")[0] == values[3]);

  vector(reparsed, "Y")[0] = 42.25;
  const auto modified = directory / "s-modified";
  reparsed.write(modified);
  assert(vector(eqmdsk::SFile(modified), "Y")[0] == 42.25);
}

void test_write_validation(const std::filesystem::path& directory) {
  const auto source = directory / "s-validation";
  const auto target = directory / "s-validation-output";
  write_bytes(source, "1 2 3 4\n");

  {
    eqmdsk::SFile file(source);
    vector(file, "DX").conservativeResize(2);
    assert_throws<eqmdsk::ValidationError>([&] { file.write(target); });
  }
  {
    eqmdsk::SFile file(source);
    vector(file, "DY")[0] = std::numeric_limits<double>::infinity();
    assert_throws<eqmdsk::ValidationError>([&] { file.write(target); });
  }
  {
    eqmdsk::SFile file(source);
    file.fields().set("X", 1.0);
    assert_throws<eqmdsk::ValidationError>([&] { file.write(target); });
  }
}

void test_deferred_write_failure(const std::filesystem::path& directory) {
  const std::filesystem::path full_device("/dev/full");
  if (!std::filesystem::exists(full_device)) {
    return;
  }
  const auto source = directory / "s-write-error";
  write_bytes(source, "1 2 3 4\n");
  const eqmdsk::SFile file(source);
  assert_throws<eqmdsk::IOError>([&] { file.write(full_device); });
}

}  // namespace

int main() {
  const TemporaryDirectory temporary;
  test_without_titles(temporary.path());
  test_three_titles_and_fortran_exponents(temporary.path());
  test_interstitial_text_and_roundtrip(temporary.path());
  test_bad_columns(temporary.path());
  test_utf8_diagnostic_paths(temporary.path());
  test_precision_modification_and_default_write(temporary.path());
  test_write_validation(temporary.path());
  test_deferred_write_failure(temporary.path());
}
